TASK-46 Add database functionality

KirSanich opened the pull request.

Commit:
TASK-46 Complex entity configuration

Настройка оставшихся сущностей

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/Account.kt:
```diff
@@ -0,0 +1,42 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    val id: Long = 0
```
в качестве айди предлагаю использовать UUID - сто процентов уникальное значение, и никогда не нужно забивать голову тем, какое же значение надо следующим подставлять

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/Account.kt:
```diff
@@ -0,0 +1,42 @@
+    @Column(name = "password")
+    var password: String = "",
+
+    @ManyToOne(cascade = [CascadeType.REFRESH], fetch = FetchType.LAZY)
```
для ManyToOne eager - дефолтный, и как правило проблемы в этом нет, и кажется даже удобнее - лишний раз не надо ходить за ролью аккаунта, а это наверняка часто используемое поле
можно оставлять его как есть

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/Account.kt:
```diff
@@ -0,0 +1,42 @@
+    @Column(name = "id")
+    val id: Long = 0
+
+    override fun equals(other: Any?): Boolean {
```
чтобы вот такие штуки каждый раз не таскать, можно завести например базовый класс 


```

@MappedSuperclass
abstract class BaseEntity<T : Serializable> {
    @get:Transient
    protected abstract val id: T?

    override fun equals(other: Any?): Boolean {
        if (null == other) return false
        if (this === other) return true
        if (javaClass != ProxyUtils.getUserClass(other)) return false

        val that = other as BaseEntity<*>
        return Objects.equals(id, that.id)
    }

    override fun hashCode(): Int {
        var hashCode = 17
        hashCode += when (id) {
            null -> 0
            else -> id.hashCode() * 31
        }
        return hashCode
    }

    override fun toString(): String {
        return "${javaClass.canonicalName}(id=$id)"
    }
}
```

```
@MappedSuperclass
abstract class UUIDBasedEntity(
    @Id
    public override val id: UUID
) : BaseEntity<UUID>()
```

```
@Entity
@Table(name = "cd_admin_user")
class AdminUserEntity(
    id: UUID,
): UUIDBasedEntity(id) {
}
```

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/Event.kt:
```diff
@@ -0,0 +1,61 @@
+    val eventInitiator: SurfEmployee = SurfEmployee(),
+
+    @ManyToMany(cascade = [CascadeType.REFRESH], fetch = FetchType.LAZY, mappedBy = "events")
+    val candidates: List<Candidate> = mutableListOf(),
```
как правило, обратные коллекции - это боль, они редко нужны в явном виде, и чаще приносят много неприятностей с LazyInitializationException, или если она вытянулась, то за собой половину базы потянет

мой фаворит среди альтернатив - это условно метод репозитория getCandidatesByEvent(event: Event)
мы всегда вызовем его только когда нам явно потребуется список кандидатов, и всегда будем уверены, что с этим проблем у нас не будет

soup1ck replied:
Очень полезный коммент, не задумывался об этом, кстати, в  м2м лучше использовать Set, так как хибер не генерит лишние запросы при модификации коллекций.

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/EventType.kt:
```diff
@@ -0,0 +1,38 @@
+    var description: String = "",
+
+    @OneToMany(cascade = [CascadeType.REFRESH], fetch = FetchType.LAZY, mappedBy = "eventType")
+    val events: List<Event> = emptyList(),
```
особенно обратные коллекции опасны у сущностей-справочников: у них вариантов значений всегда немного, а если такая коллекция вдруг проинициализируется, то для случая трех возможных значений справочников в среднем выгружается половина всех эвентов из бд

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/Question.kt:
```diff
@@ -0,0 +1,52 @@
+class Question(
+
+    @Column(name = "question")
+    var description: String = "",
```
а есть смысл в инициализации полей? пусть об этом заботится либо сервис (если происходит создание какой-нибудь сущности с заполнением полей), либо автор скрипта в бд (для случая questions скорее всего это будет ваша миграция, а явного создания этих объектов через код скорее всего и не будет)

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/entity/Question.kt:
```diff
@@ -0,0 +1,52 @@
+    @JoinTable(name = "questions_answers",
+        joinColumns = [JoinColumn(name = "answer_id")],
+        inverseJoinColumns = [JoinColumn(name = "question_id")])
+    val answers: List<Answer> = mutableListOf(),
```
с m2m ситуация аналогична: вероятнее всего никогда не потребуется знать все ответы этого вопроса, а эта штука заменится методом репозитория с кастомным запросом (в который наверняка еще добавятся нужные фильтры):

```
select a.* from answers a join question_answers qa on qa.answer_id = a.answer_id 
where qa.question_id = ? 
-- and ....
```

NikitaYakuntsev commented on core/src/main/resources/db/migration/V1.0.0__initial_script.sql:
```diff
@@ -0,0 +1,158 @@
+
+create table roles
```
неплохая конструкция для идемпотентности: create table if not exists - не уронит приложение на старте и пропустит запрос, если кто-то по ошибке почистил у себя таблицу миграций

NikitaYakuntsev commented on core/src/main/resources/db/migration/V1.0.0__initial_script.sql:
```diff
@@ -0,0 +1,158 @@
+    id       bigint primary key generated always as identity,
+    email    text not null unique,
+    password text not null,
+    role_id  bigint references roles (id)
```
хороший тон - давать имена констреинтам
вот эту штуку сгенерила идея (формат: соурс-таблица_таргет-таблица), когда я там накидал табличку для бд:
```
    user_status_id             uuid                not null
        constraint cd_user_user_status_fk
            references cd_user_status,
```

Commit:
TASK-46 Refactor entities

Исправление структуры классов сущностей, а также модификация скрипта миграции

Commit:
TASK-46 Fix constraint issues

Исправление неполадок с ограничениями

Commit:
TASK-46 Add repository support

Добавление репозиториев-заглушек, небольшой рефарктонг файлов конфигурации докера и приложения
